fix optional wait wrongly treated as false

yiwu0b11 opened the pull request:
When an optional bool in passed as optional, `isStaticallyPresent()` will actually return true.
By changing wait from `asValue` to `asAddr` and add some check can solve this.
Fixes: https://github.com/llvm/llvm-project/issues/77984

yiwu0b11 committed:
fix optional wait wrongly treated as false

When a optional bool in passed as optional, isStaticallyPresent() will actually return true.
By changing wait from asValue to asAddr and add some check can solve this.

yiwu0b11 committed:
clang-format

yi-wu-arm commented:
When testing with
```fortran
program bug
  call inner("sleep 2 && echo hi")
  print *, "Direct printout line"
contains
  subroutine inner(cmd, wait, exitstatarg, cmdstatarg, cmdmsgarg)
    character(20) :: cmd
    logical, optional :: wait
    integer, optional :: exitstatarg, cmdstatarg
    character(128), optional :: cmdmsgarg
    integer :: localvar
    call execute_command_line(cmd, wait, exitstatarg, localvar, cmdmsgarg)
  end subroutine
end
```
The console print out would be:
```
```
And when set wait to false,
```
```

yiwu0b11 committed:
remove debugging line

kiranchandramohan commented on flang/test/Lower/Intrinsics/execute_command_line.f90:
```diff
@@ -39,7 +46,7 @@ end subroutine all_args
 CHARACTER(30) :: command
 call execute_command_line(command)
-! CHECK-NEXT:     %c41_i32 = arith.constant 41 : i32 
+! CHECK-NEXT:     %c48_i32 = arith.constant 48 : i32 
```
Using the name of the constant might be prone to error. Capture this in a value and use that value for checking.

yi-wu-arm replied:
done

kiranchandramohan commented on flang/lib/Optimizer/Builder/IntrinsicCall.cpp:
```diff
@@ -2913,7 +2913,8 @@ IntrinsicLibrary::genEoshift(mlir::Type resultType,
     llvm::ArrayRef<fir::ExtendedValue> args) {
-  assert(args.size() == 5);
+  // Optional KIND argument.
+  assert(args.size() >= 1);
```
There is no KIND argument for `execute_command_line`. Is there a typo?

kiranchandramohan commented on flang/lib/Optimizer/Builder/IntrinsicCall.cpp:
```diff
@@ -2925,9 +2926,30 @@ void IntrinsicLibrary::genExecuteCommandLine(
+  if (isStaticallyAbsent(wait)) {
+    waitBool = builder.createBool(loc, true);
+  } else {
+    mlir::Type i1Ty = builder.getI1Type();
+    mlir::Value waitAddr = fir::getBase(wait);
+    mlir::Value waitIsPresentAtRuntime =
+        builder.genIsNotNullAddr(loc, waitAddr);
+    waitBool = builder
+                   .genIfOp(loc, {i1Ty}, waitIsPresentAtRuntime,
+                            /*withElseRegion=*/true)
+                   .genThen([&]() {
+                     auto waitLoad = builder.create<fir::LoadOp>(loc, waitAddr);
+                     mlir::Value cast =
+                         builder.createConvert(loc, i1Ty, waitLoad);
+                     builder.create<fir::ResultOp>(loc, cast);
+                   })
+                   .genElse([&]() {
+                     mlir::Value trueVal = builder.createBool(loc, true);
+                     builder.create<fir::ResultOp>(loc, trueVal);
+                   })
+                   .getResults()[0];
+  }
```
Is there similar handling for any other intrinsic?

luporl replied:
https://reviews.llvm.org/D123388 explains the logic behind this code.

kiranchandramohan replied:
Thanks for the pointers @yi-wu-arm @luporl

yiwu0b11 committed:
assertion and test fixes

yiwu0b11 committed:
test fixes and use variable to capture constant

psteinfeld reviewed (commented):
All builds and tests correctly and looks good to me.

You might want to wait for Kiran's approval before merging.

yiwu0b11 committed:
change assert back to args.size()==5

it turns out that it doesn't matter if the optional input are pass in or not, when compiling it is 5

kiranchandramohan reviewed (approved):
LGTM.
